sdk/storage_blobs/src/blob_client.rs

```diff
@@ -41,19 +47,91 @@ impl BlobClient {
             container_name: container_name,
             blob_name: blob_name,
             url: Url::parse(&blob_url).expect("Something went wrong with URL parsing!"),
-            pipeline: BlobClient::build_pipeline(credential, options.client_options),
+            pipeline: BlobClient::<Unset>::build_pipeline(credential, options.client_options),
+            state: Unset,
+        }
+    }
+
+    // We can make this alot better by adding the Default implements
+    fn as_block_blob(self) -> BlobClient<Block> {
+        BlobClient {
+            account_name: self.account_name,
+            credential: self.credential,
+            container_name: self.container_name,
+            blob_name: self.blob_name,
+            url: self.url,
+            pipeline: self.pipeline,
+            state: Block,
+        }
+    }
+
+    fn as_append_blob(self) -> BlobClient<Append> {
+        BlobClient {
+            account_name: self.account_name,
+            credential: self.credential,
+            container_name: self.container_name,
+            blob_name: self.blob_name,
+            url: self.url,
+            pipeline: self.pipeline,
+            state: Append,
+        }
+    }
+
+    fn as_page_blob(self) -> BlobClient<Page> {
+        BlobClient {
+            account_name: self.account_name,
+            credential: self.credential,
+            container_name: self.container_name,
+            blob_name: self.blob_name,
+            url: self.url,
+            pipeline: self.pipeline,
+            state: Page,
         }
     }
 
     // This will handle appending container and blob name
     fn build_blob_url(base_url: &str, container_name: &str, blob_name: &str) -> String {
         base_url.to_owned() + container_name + "/" + blob_name
     }
+}
+
+impl<T: BlobKind> BlobClient<T> {
+    pub async fn upload_blob(&self, data: Bytes, blob_type: Option<String>) -> Result<Response> {
```

Comment:
I realized this while working on the testcase since the content length was 0 in the upload request response, but after doing a `get_blob_properties` the content length was correct (11). So currently this is doing the initial commit + uploading the data when those are usually two discrete steps (create 0-content blob with correct type then data)